bee-ledger/src/workers/consensus/state.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    types::{Balance, Unspent},
+    workers::{
+        error::Error,
+        storage::{self, StorageBackend},
+    },
+};
+
+use bee_message::{
+    address::Address,
+    constants::IOTA_SUPPLY,
+    output::{self, dust_outputs_max},
+};
+use bee_storage::access::AsStream;
+
+use futures::StreamExt;
+
+async fn validate_ledger_unspent_state<B: StorageBackend>(storage: &B, treasury: u64) -> Result<(), Error> {
+    let mut supply: u64 = 0;
+    let mut stream = AsStream::<Unspent, ()>::stream(storage)
+        .await
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+
+    while let Some((output_id, _)) = stream.next().await {
+        let output = storage::fetch_output(storage, &*output_id)
+            .await?
+            .ok_or(Error::MissingUnspentOutput(output_id))?;
+
+        let amount = match output.inner() {
+            output::Output::SignatureLockedSingle(output) => output.amount(),
+            output::Output::SignatureLockedDustAllowance(output) => output.amount(),
+            output => return Err(Error::UnsupportedOutputKind(output.kind())),
+        };
+
+        supply = supply
+            .checked_add(amount)
+            .ok_or_else(|| Error::LedgerStateOverflow(supply as u128 + amount as u128))?;
+    }
+
+    if supply
+        .checked_add(treasury)
+        .ok_or_else(|| Error::LedgerStateOverflow(supply as u128 + treasury as u128))?
+        != IOTA_SUPPLY
+    {
+        return Err(Error::InvalidLedgerUnspentState(supply));
+    }
+
+    Ok(())
```

Comment:
I have no problems with early returns tbh, but maybe you want to consider rewriting this to :
```rust
if .... != IOTA_SUPPLY {
    Err(Error::InvalidLedgerUnspentState(supply))
} else {
    Ok(())
}
```
because this is the essential check of this function. Same in the cases down below.